cmd/portal/cmd/cmdinternal/migrate_resources_set_default_logo_height.go

```diff
@@ -0,0 +1,180 @@
+package cmdinternal
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	portalcmd "github.com/authgear/authgear-server/cmd/portal/cmd"
+	"github.com/authgear/authgear-server/cmd/portal/internal"
+	"github.com/authgear/authgear-server/pkg/lib/theme"
+)
+
+var cmdInternalMigrateSetDefaultLogoHeight = &cobra.Command{
+	Use:   "migrate-set-default-logo-height",
+	Short: "Set default logo height for apps which have logos and do not have logo height customized", // more context in DEV-2126
+	RunE: func(cmd *cobra.Command, args []string) error {
+		binder := portalcmd.GetBinder()
+
+		dbURL, err := binder.GetRequiredString(cmd, portalcmd.ArgDatabaseURL)
+		if err != nil {
+			return err
+		}
+
+		dbSchema, err := binder.GetRequiredString(cmd, portalcmd.ArgDatabaseSchema)
+		if err != nil {
+			return err
+		}
+
+		internal.MigrateResources(&internal.MigrateResourcesOptions{
+			DatabaseURL:            dbURL,
+			DatabaseSchema:         dbSchema,
+			UpdateConfigSourceFunc: migrateSetDefaultLogoHeight,
+			DryRun:                 &MigrateResourcesDryRun,
+		})
+
+		return nil
+	},
+}
+
+func migrateSetDefaultLogoHeight(appID string, configSourceData map[string]string, dryRun bool) error {
+	cfg, err := parseLogoHeightConfigSource(configSourceData)
+	if err != nil {
+		return err
+	}
+
+	hasLightLogo := cfg.LightLogo != nil
+	hasDarkLogo := cfg.DarkLogo != nil
+
+	if !hasLightLogo && !hasDarkLogo {
+		log.Printf("Skipping app (%s) because it does not have logo set", appID)
+		return nil
+	}
+
+	if dryRun {
+		log.Printf("Converting app (%s)", appID)
+	}
+
+	var lightThemeCSSMigrated []byte
+	var darkThemeCSSMigrated []byte
+	var lightThemeCSSAlreadySet bool
+	var darkThemeCSSAlreadySet bool
+	if hasLightLogo {
+		r := bytes.NewReader(cfg.LightThemeCSS.DecodedData)
+		lightThemeCSSMigrated, lightThemeCSSAlreadySet, err = theme.MigrateSetDefaultLogoHeight(r)
+		if lightThemeCSSAlreadySet {
+			log.Printf("Skipping light theme css of app (%s) because it already has logo height set", appID)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to migrate %v: %w", cfg.LightThemeCSS.OriginalPath, err)
+		}
```

Comment:
The function MigrateSetDefaultLogoHeight does not look natural. In golang, the convention of handling the result of a function that can return error is check the error first. But you check the 2nd return value first.

I suggest we have

1. A function to take `configSourceData map[string]string`, and a regexp pattern, and tell whether a path matching the pattern exists. Then we can use this function to check if a light logo or a dark logo is set. `func check(pattern *regexp.Regexp, configSourceData map[string]string) ([]string, bool)` It returns a list of path that matches the regexp. If the slice is non-empty, then 2nd return value is true, otherwise it is false.
2. A function to take the CSS string, and tell if `--brand-logo__height` is set. `func checkDeclarationInSelector(cssString string, selector string, declarationProperty string) bool`. `checkDeclarationInSelector("", ":root", "--brand-logo__height")`  
3. A function to take the CSS string, and add / replace `--brand-logo__height` with the specified value. `func setDeclarationInSelector(cssString string, selector string, declaration declaration) (newCSSString, outcome outcome)` where outcome is either `add` or `replace`.

Reply:
> A function to take the CSS string, and add / replace

Sor missed this; Why would we want to `replace`? Should it be `add/skip` instead?

Reply:
OK. Then rename the function to `AddDeclarationInSelectorIfNotPresentAlready(cssString string, selector string, declaration declaration) (newCSS string, added bool)`. If added, then `added` is true.

Reply:
You can make `declaration` public.